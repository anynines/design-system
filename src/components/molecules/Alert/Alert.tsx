import React from 'react'
import styled from 'styled-components'

import { Icon } from '../../atoms/Icon/Icon'

export type AlertType = 'success' | 'error' | 'warning'

export interface AlertProps {
  title: string
  description: string
  onClose: () => void
  type: AlertType
  className?: string
}

interface AlertStyledProps {
  type: AlertType
}

export const Alert: React.FC<AlertProps> = ({
  type = 'success',
  className = 'AlertWrapper',
  title,
  description,
  onClose
}) => {
  if (!title || !description) return null
  return (
    <AlertWrapper type={type} className={className}>
      <div className='alert-icon'>
        <Icon icon='info' />
      </div>
      <div className='alert-info'>
        <h3>{title}</h3>
        <p>{description}</p>
        <button onClick={onClose} type='button' className='alert-button'>
          <Icon icon='close' />
        </button>
      </div>
    </AlertWrapper>
  )
}

const AlertWrapper = styled.div<AlertStyledProps>`
  --position: 2rem;
  --width: 22rem;
  --z-index: 90;
  --bg-color: ${(props): string => { return props.theme.colors[props.type] }};

  display: flex;
  flex-direction: row;
  position: fixed;
  top: var(--position);
  right: var(--position);
  z-index: var(--z-index);
  width: var(--width);
  min-height: 4.5rem;
  font-size: var(--text-lg);
  border-radius: var(--radius);
  box-shadow: 0 0 .5em rgba(0,0,0,0.15);

  .alert-icon {
    display: flex;
    justify-content: center;
    align-items: center;
    background-color: var(--bg-color);
    padding: var(--space-md);
    color: var(--color-white-fix);
    border-bottom-left-radius: .5em;
    border-top-left-radius: .5em;
    filter: brightness(90%);
    flex: 1;
    max-width: 4.5rem;

    svg {
      width: 1.5em;
      height: 1.5em;
    }
  }

  .alert-info {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: flex-start;
    position: relative;
    background-color: var(--bg-color);
    padding: var(--space-md) var(--space-lg);
    color: var(--color-white-fix);
    border-bottom-right-radius: .5em;
    border-top-right-radius: .5em;
    flex: 3;

    h3 {
      margin: var(--space-sm) 0 0;
      color: var(--color-white-fix);
    }

    P {
      margin: var(--space-md) 0 var(--space-sm);
      line-height: 1.2;
    }
  }

  .alert-button{
    display: flex;
    justify-content: center;
    align-items: center;
    position: absolute;
    top: var(--space-sm);
    right: var(--space-sm);
    border: 0;
    background-color: transparent;
    cursor: pointer;
    min-width: 2em;
    min-height: 2em;
    color: var(--color-white-fix);
    outline: none;

    svg {
      width: 1em;
      height: 1em;
    }
  }
`
