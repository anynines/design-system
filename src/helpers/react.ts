import React from 'react'

export const onClickOutsideHook = (
  ref: React.RefObject<HTMLDivElement> | null,
  callback: () => void
): void => {
  // Hook get from https://stackoverflow.com/a/42234988/8583669

  // eslint-disable-next-line react-hooks/rules-of-hooks
  React.useEffect(() => {
    const handleClickOutside = (event: MouseEvent): void => {
      if (ref?.current && !(ref.current.contains(event.target as Node))) {
        callback()
      }
    }
    // Bind the event listener
    document.addEventListener('mousedown', handleClickOutside)
    return (): void => {
      // Unbind the event listener on clean up
      document.removeEventListener('mousedown', handleClickOutside)
    }
  }, [callback, ref])
}
