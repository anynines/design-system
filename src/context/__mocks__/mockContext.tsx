import React, { ReactNode } from 'react'
import { ThemeProvider } from 'styled-components'
import { render } from '@testing-library/react'

import { GlobalStyle } from '../../designSystemInstance/theme/GlobalStyle'

// D E F A U L T  T H E M E
import { defaultTheme } from '../../designSystemInstance/theme/defaultTheme'

export function renderWithProviders(children: ReactNode) {
  return render(
    <ThemeProvider theme={defaultTheme}>
      <GlobalStyle />
      {children}
    </ThemeProvider>
  )
}
