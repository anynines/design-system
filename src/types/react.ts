import { AriaAttributes, DOMAttributes } from 'react'

export type ReactHtmlElementProps<T extends AriaAttributes | DOMAttributes<T>>
  = React.DetailedHTMLProps<React.HTMLAttributes<T>, T>

export type DefaultComponentProps = Pick<ReactHtmlElementProps<HTMLDivElement>, 'className' | 'style'>
