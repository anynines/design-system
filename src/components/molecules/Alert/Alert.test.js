import React from 'react'
import { fireEvent } from '@testing-library/react'
import 'jest-styled-components'

import { Alert } from './Alert'
import { renderWithProviders } from '../../../context/__mocks__/mockContext'

describe('Alert component', () => {
  let component
  let props
  let mockFunction
  let title
  let description
  let type

  beforeEach(() => {
    mockFunction = jest.fn()
    title = 'Test Title'
    description = 'Test Description'
    type = 'success'
    props = {
      title,
      description,
      type,
      onClose: mockFunction
    }

    component = renderWithProviders(<Alert {...props} />)
  })

  describe('snapshot', () => {
    it('should match snapshot', () => {
      const { asFragment } = component
      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('rendering', () => {
    it('should not render for invalid props', () => {
      props = {
        ...props,
        ...{
          title: '',
          description: ''
        }
      }
      component = renderWithProviders(<Alert {...props} />)
      const { container } = component
      expect(container.firstChild).toBeNull()
    })
  })

  describe('event', () => {
    it('should invoke event when close icon is clicked', () => {
      const { container } = component
      const elem = container.querySelector('.alert-button')
      fireEvent.click(elem)
      expect(mockFunction).toHaveBeenCalled()
    })
  })

  describe('texts', () => {
    let container

    beforeEach(() => {
      container = component.container
    })

    it('should print correct description', () => {
      const h3 = container.querySelector('h3')
      expect(h3.textContent).toBe(title)
    })

    it('should print correct title', () => {
      const para = container.querySelector('p')
      expect(para.textContent).toBe(description)
    })
  })
})
