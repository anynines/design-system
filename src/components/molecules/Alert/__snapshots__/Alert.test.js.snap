// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert component snapshot should match snapshot 1`] = `
<DocumentFragment>
  .c1 {
  --icon-size: 1rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: var(--icon-size);
  height: var(--icon-size);
}

.c1 svg {
  width: 100%;
  height: 100%;
  max-width: var(--icon-size);
  max-height: var(--icon-size);
}

.c1.md {
  width: calc(var(--icon-size) * 2);
  height: calc(var(--icon-size) * 2);
}

.c1.md svg {
  max-width: calc(var(--icon-size) * 2);
  max-height: calc(var(--icon-size) * 2);
}

.c1.lg {
  width: calc(var(--icon-size) * 4);
  height: calc(var(--icon-size) * 4);
}

.c1.lg svg {
  max-width: calc(var(--icon-size) * 4);
  max-height: calc(var(--icon-size) * 4);
}

.c0 {
  --width: 22rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  z-index: var(--z-index);
  width: var(--width);
  min-height: 4.5rem;
  border-radius: var(--radius);
  box-shadow: 0 0 .5em rgba(0,0,0,0.15);
}

.c0 .alert-icon {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: var(--bg-color);
  padding: var(--space-md);
  color: var(--color-white-fix);
  border-bottom-left-radius: .5em;
  border-top-left-radius: .5em;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  max-width: 4.5rem;
}

.c0 .alert-icon svg {
  width: 1.5em;
  height: 1.5em;
  max-width: 1.5em;
  max-height: 1.5em;
}

.c0 .alert-info {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  position: relative;
  background-color: var(--bg-color);
  padding: var(--space-sm) var(--space-md);
  border-bottom-right-radius: .5em;
  border-top-right-radius: .5em;
  -webkit-flex: 3;
  -ms-flex: 3;
  flex: 3;
}

.c0 .alert-info h4,
.c0 .alert-info p {
  color: var(--color-white-fix);
}

.c0 .alert-info span {
  font-weight: var(--font-weight-bd);
}

.c0 .alert-info P {
  margin-top: var(--space-xs);
  margin-bottom: 0;
  line-height: 1.2;
}

.c0 .alert-button {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: absolute;
  top: var(--space-sm);
  right: var(--space-sm);
  border: 0;
  background-color: transparent;
  cursor: pointer;
  min-width: 2em;
  min-height: 2em;
  color: var(--color-white-fix);
  outline: none;
}

.c0 .alert-button svg {
  width: 1em;
  height: 1em;
}

.c0.notice {
  background-color: var(--color-notice);
}

.c0.notice .alert-icon {
  background-color: var(--color-notice-dark);
}

.c0.success {
  background-color: var(--color-success);
}

.c0.success .alert-icon {
  background-color: var(--color-success-dark);
}

.c0.warning {
  background-color: var(--color-warning);
}

.c0.warning .alert-icon {
  background-color: var(--color-warning-dark);
}

.c0.error {
  background-color: var(--color-error);
}

.c0.error .alert-icon {
  background-color: var(--color-error-dark);
}

<div
    class="c0 alert-wrapper undefined success"
  >
    <div
      class="alert-icon"
    >
      <i
        class="c1 icon undefined md"
      >
        <svg
          height="24.707"
          viewBox="0 0 24.707 24.707"
          width="24.707"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.353,0A12.353,12.353,0,1,0,24.707,12.353,12.367,12.367,0,0,0,12.353,0Zm.8,19.683c-.587.1-1.755.342-2.348.391a1.445,1.445,0,0,1-1.264-.658A1.544,1.544,0,0,1,9.358,18l2.335-6.421H9.265a2.9,2.9,0,0,1,2.285-2.7A8.939,8.939,0,0,1,13.9,8.493a1.862,1.862,0,0,1,1.264.658,1.544,1.544,0,0,1,.187,1.414l-2.335,6.421h2.427a2.733,2.733,0,0,1-2.284,2.7ZM13.9,7.721a1.544,1.544,0,1,1,1.544-1.544A1.544,1.544,0,0,1,13.9,7.721Z"
            fill="currentColor"
            id="info"
          />
        </svg>
      </i>
    </div>
    <div
      class="alert-info"
    >
      <span>
        Test Title
      </span>
      <p>
        Test Description
      </p>
      <button
        class="alert-button"
        type="button"
      >
        <i
          class="c1 icon undefined sm"
        >
          <svg
            height="10.852"
            viewBox="0 0 10.852 10.852"
            width="10.852"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              data-name="Gruppe 2951"
              id="Gruppe_2951"
              transform="translate(-2008.124 592.325) rotate(-45)"
            >
              <rect
                data-name="Rechteck 507"
                fill="currentColor"
                height="2.79"
                id="Rechteck_507"
                rx="1.395"
                transform="translate(1832.517 1007.4)"
                width="12.557"
              />
              <rect
                data-name="Rechteck 508"
                fill="currentColor"
                height="2.79"
                id="Rechteck_508"
                rx="1.395"
                transform="translate(1837.4 1015.073) rotate(-90)"
                width="12.557"
              />
            </g>
          </svg>
        </i>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
